### PR TITLE
respond to ADD_TRANSACTION if it looks like a key purchase

### PR DESCRIPTION
**Pull request by cellog**
# Description

This is the last piece of the fix for #3246. When locksmith passes a pending transaction to the paywall, and the user has refreshed the page, we don't have enough information to do anything with it. `web3Service.getTransaction` will return null, and the page displays "your key is expired" or "purchase a key."

This PR adds some "stuffing" of the action if it looks like a key purchase (from us to the lock on the page). The stuffing is the transaction type, lock and key, which is enough for the UI to respond to it. We also have to make sure the transaction shows up in the key transaction list, which is at the bottom of the file.

Because of this change, it now means we handle `ADD_TRANSACTION`, `NEW_TRANSACTION`, and `UPDATE_TRANSACTION` the same way, so the code refactors the bottom of the file to do this, abstracting the small differences between the structure of each action.

After this PR, a release of unlock-js followed by upgrade to paywall will fix the bug.

# Issues

Refs #3246 

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [X] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [X] My code follows the style guidelines of this project, including naming conventions
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [X] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by julien51** [comment]
> This PR adds some "stuffing" of the action if it looks like a key purchase (from us to the lock on the page).

Well that is weird because `getTransaction` should NOT return `null`? If the transaction is pending it should actually return some info from the mempool indicating that the transaction has not been mined yet. Web3Service actually has the logic to look at the `data` to identify that it is for sure an actual keey purchase.

Maybe we broke Web3Service?

File: paywall/src/middlewares/web3Middleware.js [hunk 1]
```diff
@@ -105,6 +106,24 @@ const web3Middleware = config => ({ getState, dispatch }) => {
...
+            location: { pathname, hash },
+          },
+        } = getState()
+        const { lockAddress } = lockRoute(pathname + hash)
+        const accountAddress = account && account.address
+        const { to, from } = action.transaction
+        if (from === accountAddress && to === lockAddress) {
+          // this is probably a key purchase transaction, so let's make sure we know it is in the UI
+          // pending key transactions will never see this detail otherwise
+          action.transaction.type = TRANSACTION_TYPES.KEY_PURCHASE
```
Comment (julien51) [inline paywall/src/middlewares/web3Middleware.js]: Nah, this is not right we should not do guesswork here. Web3Service **has** the ability to retrieve that info (and used to!), so let's make sure it does.

**Comment by cellog** [comment]
Julien the reason that it has to be here is because the pending transaction cannot be retrieved from the read only provider because it is still pending. We have defaults for this exact case, but the defaults don't exist when the user refreshes. It could be fixed by caching the defaults in localStorage, but that is riskier imo

**Comment by cellog** [comment]
As an alternative, we can leave this unfixed until the work I'm doing for the ad remover paywall is live, because that *will* in fact cache the transaction value, but safely

**Pull request closed without merging**